plugins/aggregate/aggregate.cpp

```diff
@@ -247,21 +256,152 @@ struct aggregation {
         batch = set_column_result.MoveValueUnsafe();
       }
     }
-    // Third and last, aggregate the batch eagerly.
-    return aggregate({std::move(batch)});
+    // Iterate over the record batches row-wise and select slices that group
+    // into the same bucket as large as possible, adding them into the buckets
+    // accumulator state one slice at a time.
+    VAST_ASSERT(batch->num_rows() >= 1);
+    int start = 0;
+    int next_start = 0;
+    auto next_bucket = try_emplace_bucket(batch, start);
+    for (auto bucket = next_bucket; next_bucket != buckets_.end();
+         bucket = next_bucket) {
+      start = next_start;
+      do {
+        next_bucket = try_emplace_bucket(batch, ++next_start);
+      } while (bucket == next_bucket);
+      VAST_ASSERT(bucket->second.size() == actions_.size());
+      for (int column = 0; column < batch->num_columns(); ++column) {
+        auto f =
+          [&]<class Array>([[maybe_unused]] const Array& array) -> caf::error {
+          if constexpr (!std::is_base_of_v<arrow::PrimitiveArray, Array>) {
+            return caf::make_error(
+              ec::invalid_configuration,
+              fmt::format("aggregate transform step cannot handle "
+                          "non-primitive field {}",
+                          batch->schema()->field(column)->ToString()));
+          } else {
+            using traits = arrow::TypeTraits<typename Array::TypeClass>;
```

Comment:
The trait needs to be specialized for our extension types, because otherwise there's no way to get a scalar from the Array type. But that can all come once the experimental stuff goes live.